### PR TITLE
Improve the performance of the index script

### PR DESCRIPTION
**Pull request by amorde**
This brings the runtime down from ~4 minutes to ~1 minute on my machine. Performance will likely vary on Netlify

### CODE REVIEW

**Comment by amorde** [comment]
cc @igor-makarov

**Comment by igor-makarov** [comment]
Do you think you could reuse `concurrent-ruby` as we already have one concurrency library?

**Comment by amorde** [comment]
We actually have both `parallel` _and_ concurrent ruby, I only added it explicitly so its not implicit. I'm familiar with `Parallel` which is why I used it

**Comment by igor-makarov** [comment]
Gotcha! Will look at it closer tomorrow morning 👍🏻

File: Scripts/create_pods_and_versions_index.rb [hunk 1]
```diff
@@ -2,47 +2,150 @@
 # frozen_string_literal: true
 
 require 'cocoapods'
+require 'parallel'
 
-source = Pod::Source.new('.')
+def parallel?
```
Comment (igor-makarov) [inline Scripts/create_pods_and_versions_index.rb:7]: Is there any reason for this configuration?

File: Scripts/create_pods_and_versions_index.rb [hunk 2]
```diff
@@ -2,47 +2,150 @@
...
+  ENV['COCOAPODS_PARALLEL']
+end
+
+def netlify?
```
Comment (igor-makarov) [inline Scripts/create_pods_and_versions_index.rb:11]: Is there any reason for this configuration?

File: Scripts/create_pods_and_versions_index.rb [hunk 3]
```diff
@@ -2,47 +2,150 @@
...
+      io.puts msg
+    else
+      io ||= STDOUT
+      timestamp = Time.now.strftime('%k:%M:%S')
+      io.puts "[#{timestamp}] #{msg}"
+    end
+  end
+end
+
+module Pod
```
Comment (igor-makarov) [inline Scripts/create_pods_and_versions_index.rb:28]: Perhaps move into a separate source file?

File: Scripts/create_pods_and_versions_index.rb [hunk 4]
```diff
@@ -2,47 +2,150 @@
...
-source = Pod::Source.new('.')
+def parallel?
+  ENV['COCOAPODS_PARALLEL']
+end
+
+def netlify?
+  ENV['NETLIFY']
+end
+
+class UI
```
Comment (igor-makarov) [inline Scripts/create_pods_and_versions_index.rb:15]: Perhaps move into a separate source file?

File: Scripts/create_pods_and_versions_index.rb [hunk 5]
```diff
@@ -2,47 +2,150 @@
...
+        Pathname.glob(glob).reduce([]) do |pods, entry|
+          pods << entry.basename.to_s if entry.directory?
+          pods
+        end.sort
+      end
+    end
+  end
+end
+
+def parallel_map(iter, &block)
```
Comment (igor-makarov) [inline Scripts/create_pods_and_versions_index.rb:54]: Perhaps move into a separate source file?

**Review by igor-makarov — CHANGES_REQUESTED** [review]

**Comment by igor-makarov** [comment]
Left some specific comments regarding source structure. Looking good!

By the way, did you check that the sort order remains correct?

File: Scripts/create_pods_and_versions_index.rb (hunk 2 above)
Comment (amorde) [reply Scripts/create_pods_and_versions_index.rb:11]: I wanted timestamps on the logs to see how long each step takes, but Netlify automatically includes this in its logs so I didn't want to double-up

File: Scripts/create_pods_and_versions_index.rb (hunk 1 above)
Comment (amorde) [reply Scripts/create_pods_and_versions_index.rb:7]: wanted to make it easy to turn off parallelism if it caused issues, though I can see how that might not be worthwhile

**Comment by amorde** [comment]
Checking the order now, here's an example of the shards

`all_pods_versions_0_0_1.txt`:
```
AsciiNumberTools/0.0.2
AuroraIMUI/0.1.0/0.10.1/0.10.11/0.10.12/0.12.0/0.12.7/0.14.0/0.2.0/0.4.0/0.5.3/0.5.4/0.6.0/0.6.2/0.6.7
CGXPopoverView/0.5/0.7/0.8/1.0
CYXImagePickerManager/0.0.2/0.0.3
CommonPrivado/0.0.1
CrossSlideBar/0.0.1/0.0.11/0.0.2/0.0.3/0.0.4/0.0.5/0.0.6/0.0.7/0.0.8/0.0.9/2.0.0/2.0.1
DCTimerManager/0.0.1/0.0.2
DDMoreDebug/0.0.6
DistImagePickerView/1.0.0/1.0.1/1.0.2/1.0.3
FTLogger/1.0.0/1.0.1
FoxconnKit/0.1.0
GuidolTNSIOSPlugin/1.0.2/1.0.3/1.0.4
HJMacros/0.1.0/0.1.1/0.1.2/0.1.3/0.2.2/0.2.3/0.3.0/0.3.1/0.3.2/0.4.0/0.5.0/0.6.0/0.7.0
HapticFeedback/1.0
IGListDiffKit/4.0.0
MixRoute/1.0.0/1.0.1/1.0.2/1.0.3/1.0.4/1.0.5/1.0.6
MixfaderSDKCocoa/2.0/2.0.1/2.0.2/2.1/2.1.1
ModelCoder/1.0
SGPush/0.1
SwiftPasscodeLock/1.1.3/1.1.4/1.1.5
USeek/0.1.0/0.1.3/0.1.5/0.1.6/0.1.9/0.2.1/0.2.2/0.2.3/0.2.6/0.2.7/0.3.0/0.3.1/0.3.2/0.4.1/0.5.0
YYImage/0.8.9/0.9.0/0.9.1/0.9.2/0.9.3/0.9.4/0.9.5/1.0/1.0.1/1.0.2/1.0.3/1.0.4
lolAPIWrapper/1.0.0/1.1.0/1.1.1/1.1.2/1.1.3/1.1.4/1.1.5/1.1.7/1.1.8
```
Seems to match https://cdn.cocoapods.org/all_pods_versions_0_0_1.txt


Deprecated pods were sorted incorrectly and were also using absolute paths, will need to fix that

File: Scripts/create_pods_and_versions_index.rb (hunk 4 above)
Comment (amorde) [reply Scripts/create_pods_and_versions_index.rb:18]: Have moved the others but I think it makes sense to keep this here

File: Scripts/create_pods_and_versions_index.rb (hunk 2 above)
Comment (amorde) [reply Scripts/create_pods_and_versions_index.rb]: ended up moving it

File: Scripts/create_pods_and_versions_index.rb (hunk 4 above)
Comment (amorde) [reply Scripts/create_pods_and_versions_index.rb]: ended up moving it

**Comment by orta** [comment]
Overall, this looks good, though parallelism like most threading things is always hard to review from a read of the code.

I'd recommend merging it on a weekend to handle the case of it having differences

**Review by dnkoutso — APPROVED** [review]

**Comment by igor-makarov** [comment]
I've added the branch to Netlify's builds and it failed. I think that subsequent pushes to the branch will get built as well.

Here's the failing build, looks like a Ruby/Bundler issue: https://app.netlify.com/sites/cocoapods-cdn/deploys/5ec4cae94ba50b41bb69614b

File: Gemfile.lock [hunk 6]
```diff
@@ -1,25 +1,28 @@
 GEM
   remote: https://rubygems.org/
   specs:
-    CFPropertyList (3.0.0)
-    activesupport (4.2.11.1)
+    CFPropertyList (3.0.2)
+    activesupport (4.2.11.12)
```
Comment (igor-makarov) [inline Gemfile.lock:5]: 4.2.11.12 seems wrong @amorde

Comment (amorde) [reply Gemfile.lock:5]: Ugh we have run into this in our other repos, I don’t know why it keeps happening. @dnkoutso was the fix to manually edit the lock file?

Comment (dnkoutso) [reply Gemfile.lock:5]: yes unfortunately :(

**Comment by amorde** [comment]
Going to close and re-open since we seem to have force-pushes disabled event for non-master branches

**Pull request closed without merging**